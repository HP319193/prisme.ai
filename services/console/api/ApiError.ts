import { APIError } from "../../../packages/types";

export class ApiError extends Error implements APIError {
  public code: number;
  public error: APIError["error"];
  public details: APIError["details"];

  constructor({ error, message, details }: APIError, code: number) {
    super(message);
    this.code = code;
    this.error = error;
    this.details = details;
  }

  toString() {
    return this.error;
  }
}

export default ApiError;
