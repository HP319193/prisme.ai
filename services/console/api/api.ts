import "@prisme.ai/types";
import Fetcher from "./fetcher";

export class Api extends Fetcher {
  async me() {
    return await this.get("/me");
  }

  async signin(
    email: string,
    password: string
  ): Promise<
    Prismeai.GenericError & {
      headers: {
        ["x-prismeai-session-token"]: string;
      };
    }
  > {
    return await this.post("/login", {
      email,
      password,
    });
  }
}

export default new Api(process.env.NEXT_PUBLIC_API_HOST || "");
