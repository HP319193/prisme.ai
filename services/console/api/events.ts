import getConfig from 'next/config';
import io, { Socket } from 'socket.io-client';

const { publicRuntimeConfig } = getConfig();

export class Events {
  private client: Socket;
  constructor(workspaceId: string) {
    this.client = io(
      `${publicRuntimeConfig.API_HOST || ''}/workspaces/${workspaceId}/events`,
      {
        withCredentials: true,
      }
    );
  }

  destroy() {
    this.client.disconnect();
  }

  all(listener: (eventName: string, eventData: Prismeai.PrismeEvent) => void) {
    this.client.onAny(listener);

    return () => this.client.offAny(listener);
  }
}

export default Events;
