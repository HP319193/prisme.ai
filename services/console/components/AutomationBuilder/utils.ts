import { Schema } from '@prisme.ai/design-system';

const HAS_NO_FORM = ['all', 'conditions'];
export const instructionHasForm = (name: string, schema: Schema) => {
  return !HAS_NO_FORM.includes(name) && schema.type;
};
