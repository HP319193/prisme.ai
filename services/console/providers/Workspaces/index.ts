export * from './WorkspacesProvider';
export { default } from './WorkspacesProvider';
