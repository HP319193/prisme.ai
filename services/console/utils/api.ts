import getConfig from 'next/config';
import { Api } from '@prisme.ai/sdk';
import Storage from './Storage';
import isServerSide from './isServerSide';

const { publicRuntimeConfig } = getConfig();

function getRedirectURI() {
  try {
    return new URL(
      '/signin',
      publicRuntimeConfig.CONSOLE_URL || 'http://localhost:3000'
    ).toString();
  } catch {
    return '';
  }
}

// Optional : If INTERNAL_API_URL is specified we can use it on server side calls. Useful for on-premise deployments.
const host =
  isServerSide() && publicRuntimeConfig.INTERNAL_API_URL
    ? publicRuntimeConfig.INTERNAL_API_URL
    : publicRuntimeConfig.API_URL;

const api = new Api({
  host: host,
  oidc: {
    url: publicRuntimeConfig.OIDC_PROVIDER_URL,
    clientId: publicRuntimeConfig.OIDC_STUDIO_CLIENT_ID,
    clientIdHeader: publicRuntimeConfig.OIDC_CLIENT_ID_HEADER,
    redirectUri: getRedirectURI(),
  },
});
api.token = Storage.get('access-token');
const legacyToken = Storage.get('auth-token');
if (!api.token && legacyToken) {
  api.legacyToken = legacyToken;
}

export * from '@prisme.ai/sdk';
export default api;
