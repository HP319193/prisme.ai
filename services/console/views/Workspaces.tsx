import { Layout, Button, Text, Space, Title } from '@prisme.ai/design-system';
import Head from 'next/head';
import Link from 'next/link';
import Image from 'next/image';
import { useRouter } from 'next/router';
import { Card } from 'antd';
import { PlusOutlined } from '@ant-design/icons';
import { useCallback, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Workspace } from '../api/types';
import Header from '../components/Header';
import { useWorkspaces } from '../components/WorkspacesProvider';
import icon from '../icons/icon-workspace.svg';

export const WorkspacesView = () => {
  const { t } = useTranslation('workspaces');
  const { push } = useRouter();
  const { workspaces, create } = useWorkspaces();
  const [loading, setLoading] = useState(false);
  const workspacesList = useMemo(
    () => Array.from(workspaces.values()).filter(Boolean) as Workspace[],
    [workspaces]
  );

  const createWorkspace = useCallback(async () => {
    setLoading(true);
    const { id } = await create(t('create.defaultName'));
    push(`/workspaces/${id}`);
    setLoading(false);
  }, [create, push, t]);

  return (
    <>
      <Head>
        <title>{t('workspaces.title')}</title>
        <meta name="description" content={t('workspaces.description')} />ยง
      </Head>
      <Layout Header={<Header />}>
        <div className="!bg-blue-200 flex grow m-4 rounded">
          <div className="flex flex-wrap align-start justify-start">
            {workspacesList.map(({ name, id }) => (
              <Card
                key={id}
                className="!m-8 w-64 h-96 flex flex-col justify-between overflow-hidden"
                actions={[
                  <Link href={`/workspaces/${id}`} key="1">
                    <a>
                      <Button>{t('edit.label')}</Button>
                    </a>
                  </Link>,
                ]}
              >
                <div className="flex grow items-center justify-center flex-col text-center">
                  <Space direction="vertical">
                    <Image
                      src={icon}
                      width={48}
                      height={48}
                      className="rounded text-blue"
                      alt={name}
                    />
                    <Title level={4}>{name}</Title>
                    <Text type="grey">
                      {t('workspaces.defaultDescription')}
                    </Text>
                  </Space>
                </div>
              </Card>
            ))}
            <Card className="!m-8 w-64 h-96 flex flex-col justify-between overflow-hidden !bg-transparent !border-2">
              <div className="flex grow items-center justify-between flex-col text-center">
                <div className="flex items-center flex-col text-center mt-20">
                  <Button
                    variant="primary"
                    onClick={createWorkspace}
                    disabled={loading}
                    id="createWorkspaceButton"
                    className="!h-10 !w-10 !p-0 !flex items-center justify-center mb-5"
                  >
                    <PlusOutlined />
                  </Button>
                  <Title level={4}>
                    {t('create.label', {
                      context: workspaces.size === 0 ? 'first' : '',
                    })}
                  </Title>
                </div>
                <Text type="grey">
                  {t('create.description_sub', {
                    context: workspaces.size === 0 ? 'first' : '',
                  })}
                </Text>
              </div>
            </Card>
          </div>
        </div>
      </Layout>
    </>
  );
};

export default WorkspacesView;
