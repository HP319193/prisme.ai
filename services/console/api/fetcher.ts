import { Storage } from "../utils/Storage";
import ApiError from "./ApiError";
import HTTPError from "./HTTPError";

const headersAsObject = (headers: Headers) =>
  Array.from(headers).reduce(
    (prev, [k, v]) => ({
      ...prev,
      [k]: v,
    }),
    {}
  );

export class Fetcher {
  private host: string;

  set token(v: string | null) {
    if (v) {
      Storage.set("auth-token", v);
    } else {
      Storage.remove("auth-token");
    }
  }

  private get _token() {
    return Storage.get("auth-token");
  }

  constructor(host: string) {
    this.host = host;
  }

  private async _fetch(url: string, options: RequestInit = {}) {
    const headers: any = options.headers || {};
    if (this._token && !headers["x-prismeai-session-token"]) {
      headers["x-prismeai-session-token"] = this._token;
    }
    const res = await global.fetch(`${this.host}${url}`, {
      ...options,
      headers: {
        ...headers,
        "Content-Type": "application/json",
      },
    });

    if (!res.ok) {
      let error;
      try {
        error = new ApiError(JSON.parse(res.statusText), res.status);
      } catch (e) {
        error = new HTTPError(res.statusText, res.status);
      }
      throw error;
    }

    const response = (await res.json()) || {};
    Object.defineProperty(response, "headers", {
      value: headersAsObject(res.headers),
      configurable: false,
      enumerable: false,
      writable: false,
    });
    return response;
  }

  async get(url: string) {
    return this._fetch(url, {
      method: "GET",
    });
  }

  async post(url: string, body?: Record<string, any>) {
    return this._fetch(url, {
      method: "POST",
      body: body && JSON.stringify(body),
    });
  }

  async patch(url: string, body: Record<string, any>) {
    return this._fetch(url, {
      method: "PATCH",
      body: JSON.stringify(body),
    });
  }

  async delete(url: string, id: string) {
    return this._fetch(url, {
      method: "DELETE",
      body: JSON.stringify({ id }),
    });
  }
}

export default Fetcher;
