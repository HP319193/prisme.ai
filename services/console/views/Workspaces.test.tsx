import Workspaces from './Workspaces';
import renderer, { act } from 'react-test-renderer';
import { useWorkspaces } from '../providers/Workspaces';
import { useRouter } from 'next/router';
import { Workspace } from '../utils/api';

jest.mock('../providers/Workspaces', () => {
  const workspaces: Workspace[] = [];
  const createWorkspace = jest.fn(() => ({
    id: '43',
  }));
  const loading = new Map();
  const LoadingType = {
    New: 'New',
    List: 'List',
  };

  const fetchWorkspaces = jest.fn();

  return {
    useWorkspaces: () => ({
      workspaces,
      createWorkspace,
      loading,
      fetchWorkspaces,
    }),
    LoadingType,
  };
});
jest.mock('next/router', () => {
  const push = jest.fn();
  return {
    useRouter: () => ({
      push,
    }),
  };
});
jest.mock('next/image', () => {
  const Image = ({ src }: any) => <div />;

  return Image;
});

beforeEach(() => {
  useWorkspaces().workspaces = [];
});

it('should render empty', () => {
  const root = renderer.create(<Workspaces />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should render some workspaces', () => {
  useWorkspaces().workspaces = [
    {
      id: '1',
      name: 'foo',
      createdAt: new Date('2021-12-15'),
      createdBy: '123',
      updatedAt: new Date('2021-12-15'),
      updatedBy: '123',
    },
    {
      id: '42',
      name: 'bar',
      createdAt: new Date('2021-12-15'),
      createdBy: '123',
      updatedAt: new Date('2021-12-15'),
      updatedBy: '123',
    },
  ];
  const root = renderer.create(<Workspaces />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should create new workspace', async () => {
  const root = renderer.create(<Workspaces />);
  await act(async () => {
    const createButton = root.root.find((a) => {
      return a.props?.children?.[1]?.props?.children === 'create.label';
    });
    await createButton.props.onClick();
  });
  expect(useWorkspaces().createWorkspace).toHaveBeenCalledWith(
    'create.defaultName'
  );
  expect(useRouter().push).toHaveBeenCalledWith('/workspaces/43');
});
