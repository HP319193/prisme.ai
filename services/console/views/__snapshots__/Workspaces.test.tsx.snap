// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render empty 1`] = `
Array [
  <div
    className="mx-32 my-16"
  >
    <div
      className="bg-info px-14 py-8 rounded-[15px]"
    >
      <div
        className="text-2xl py-3 font-bold"
      >
        workspaces.welcome.title
      </div>
      <div
        className="text-prisme-darkblue"
      >
        workspaces.welcome.content
      </div>
      <div
        className="text-accent mt-4 text-sm"
      />
    </div>
    <div
      className="pt-6"
    />
    <div
      className="pt-10"
    >
      <div
        className="text-xl py-3 font-bold"
      >
        workspaces.suggestions.title
      </div>
      <div
        className="flex flex-nowrap -mx-2 sm:flex-col md:flex-row"
      >
        <div
          className="
        relative group flex flex-col h-32
        w-[100%] md:w-[calc(100%/2)] lg:w-[calc(100%/3)] xl:w-[calc(100%/4)]
        "
        >
          <button
            className="flex m-2 h-full justify-between overflow-hidden rounded border
          transition-all
          hover:-translate-y-1 hover:shadow-lg
          p-6 flex border-accent border-dashed bg-ultra-light-accent items-center !justify-start"
            onClick={[Function]}
          >
            <span
              className="flex min-w-[50px] bg-accent p-4 rounded"
            >
              <div />
            </span>
            <span
              className="flex font-bold ml-4 "
            >
              create.label
            </span>
          </button>
        </div>
        <div
          className="flex relative overflow-hidden flex-1 pb-4 -mb-4"
        >
          <div
            className="flex relative overflow-auto flex-1 pb-4 -mb-4"
            style={
              Object {
                "scrollSnapType": "x mandatory",
              }
            }
          >
            <div
              className="flex flex-1"
            />
          </div>
          <button
            className="absolute right-0 !text-light-gray focus:outline-none text-xl"
            onClick={[Function]}
            style={
              Object {
                "top": "calc(50% - 20px)",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>,
  <div
    className="absolute bottom-1 right-1 text-gray mr-1 text-[10px]"
  >
    Prisme.ai v
    6.8.0
  </div>,
]
`;

exports[`should render some workspaces 1`] = `
Array [
  <div
    className="mx-32 my-16"
  >
    <div
      className="bg-info px-14 py-8 rounded-[15px]"
    >
      <div
        className="text-2xl py-3 font-bold"
      >
        workspaces.welcome.title
      </div>
      <div
        className="text-prisme-darkblue"
      >
        workspaces.welcome.content
      </div>
      <div
        className="text-accent mt-4 text-sm"
      />
    </div>
    <div
      className="pt-6"
    />
    <div
      className="pt-10"
    >
      <div
        className="text-xl py-3 font-bold"
      >
        workspaces.suggestions.title
      </div>
      <div
        className="flex flex-nowrap -mx-2 sm:flex-col md:flex-row"
      >
        <div
          className="
        relative group flex flex-col h-32
        w-[100%] md:w-[calc(100%/2)] lg:w-[calc(100%/3)] xl:w-[calc(100%/4)]
        "
        >
          <button
            className="flex m-2 h-full justify-between overflow-hidden rounded border
          transition-all
          hover:-translate-y-1 hover:shadow-lg
          p-6 flex border-accent border-dashed bg-ultra-light-accent items-center !justify-start"
            onClick={[Function]}
          >
            <span
              className="flex min-w-[50px] bg-accent p-4 rounded"
            >
              <div />
            </span>
            <span
              className="flex font-bold ml-4 "
            >
              create.label
            </span>
          </button>
        </div>
        <div
          className="flex relative overflow-hidden flex-1 pb-4 -mb-4"
        >
          <div
            className="flex relative overflow-auto flex-1 pb-4 -mb-4"
            style={
              Object {
                "scrollSnapType": "x mandatory",
              }
            }
          >
            <div
              className="flex flex-1"
            />
          </div>
          <button
            className="absolute right-0 !text-light-gray focus:outline-none text-xl"
            onClick={[Function]}
            style={
              Object {
                "top": "calc(50% - 20px)",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>,
  <div
    className="absolute bottom-1 right-1 text-gray mr-1 text-[10px]"
  >
    Prisme.ai v
    6.8.0
  </div>,
]
`;
